File: solana_toolbox_web3/src/ToolboxIdlInstructionBlob.compute.ts
import { PublicKey } from '@solana/web3.js';
import {
  ToolboxIdlInstructionBlob,
  ToolboxIdlInstructionBlobAccount,
  ToolboxIdlInstructionBlobArg,
  ToolboxIdlInstructionBlobConst,
} from './ToolboxIdlInstructionBlob';
import { ToolboxIdlTypeFull } from './ToolboxIdlTypeFull';
import { serialize } from './ToolboxIdlTypeFull.serialize';
import { pathGetJsonValue } from './ToolboxIdlPath.json';
import { pathGetTypeFull } from './ToolboxIdlPath.type';

export type ToolboxIdlInstructionBlobComputeContext = {
  instructionProgramId: PublicKey;
  instructionPayload: any;
  instructionAddresses: Map<string, PublicKey>;
  instructionAccountsStates: Map<string, any>;
  instructionAccountsContentsTypeFull: Map<string, ToolboxIdlTypeFull>;
};

export function computeInstructionBlob(
  blob: ToolboxIdlInstructionBlob,
  context: ToolboxIdlInstructionBlobComputeContext,
): Buffer {
  return blob.traverse(computeVisitor, context, undefined);
}

const computeVisitor = {
  const: (
    self: ToolboxIdlInstructionBlobConst,
    _context: ToolboxIdlInstructionBlobComputeContext,
  ) => {
    return self.value;
  },
  arg: (
    self: ToolboxIdlInstructionBlobArg,
    context: ToolboxIdlInstructionBlobComputeContext,
  ) => {
    const value = pathGetJsonValue(self.path, context.instructionPayload);
    const data = new Array<Buffer>();
    serialize(self.typeFull, value, data, false);
    return Buffer.concat(data);
  },
  account: (
    self: ToolboxIdlInstructionBlobAccount,
    context: ToolboxIdlInstructionBlobComputeContext,
  ) => {
    if (self.path.isEmpty()) {
      throw new Error(
        'PDA Blob account path is empty (should have at least the account name)',
      );
    }
    const { first: current, rest: next } = self.path.splitFirst()!;
    const instructionAccountName = current.key();
    if (!instructionAccountName) {
      throw new Error(
        'PDA Blob account path first part should be an account name',
      );
    }
    const instructionAccountContentPath = next;
    if (instructionAccountContentPath.isEmpty()) {
      const instructionAddress = context.instructionAddresses.get(
        instructionAccountName,
      )!;
      if (!instructionAddress) {
        throw new Error(
          `Could not find address for account: ${instructionAccountName}`,
        );
      }
      return instructionAddress.toBuffer();
    }
    const instructionAccountContentState =
      context.instructionAccountsStates.get(instructionAccountName);
    if (instructionAccountContentState === undefined) {
      throw new Error(
        `Could not find state for account: ${instructionAccountName}`,
      );
    }
    const value = pathGetJsonValue(
      instructionAccountContentPath,
      instructionAccountContentState,
    );
    const data = new Array<Buffer>();
    if (self.typeFull !== undefined) {
      serialize(self.typeFull, value, data, false);
      return Buffer.concat(data);
    }
    const instructionAccountContentTypeFull =
      context.instructionAccountsContentsTypeFull.get(instructionAccountName);
    if (instructionAccountContentTypeFull === undefined) {
      throw new Error(
        `Could not find content type for account: ${instructionAccountName}`,
      );
    }
    const typeFull = pathGetTypeFull(
      instructionAccountContentPath,
      instructionAccountContentTypeFull,
    );
    serialize(typeFull, value, data, false);
    return Buffer.concat(data);
  },
};

File: solana_toolbox_idl/src/toolbox_idl_instruction_blob_compute.rs
use std::collections::HashMap;
use std::sync::Arc;

use anyhow::anyhow;
use anyhow::Context;
use anyhow::Result;
use serde_json::Value;
use solana_sdk::pubkey::Pubkey;

use crate::toolbox_idl_instruction_blob::ToolboxIdlInstructionBlob;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFull;
use crate::toolbox_idl_utils::idl_map_get_key_or_else;

impl ToolboxIdlInstructionBlob {
    pub fn try_compute(
        &self,
        instruction_payload: &Value,
        instruction_addresses: &HashMap<String, Pubkey>,
        instruction_accounts_states: &HashMap<String, Value>,
        instruction_accounts_contents_type_full: &HashMap<
            String,
            Arc<ToolboxIdlTypeFull>,
        >,
    ) -> Result<Vec<u8>> {
        let mut data = vec![];
        let prefixed = false;
        match self {
            ToolboxIdlInstructionBlob::Const {
                value, type_full, ..
            } => {
                type_full
                    .try_serialize(value, &mut data, prefixed)
                    .context("Serialize const bytes")?;
            },
            ToolboxIdlInstructionBlob::Arg {
                path, type_full, ..
            } => {
                let value = path
                    .try_get_json_value(instruction_payload)
                    .context("Extract arg value")?;
                type_full
                    .try_serialize(value, &mut data, prefixed)
                    .context("Serialize arg bytes")?;
            },
            ToolboxIdlInstructionBlob::Account { path, typing, .. } => {
                let (instruction_account_name, account_content_path) =
                    path.split_first().ok_or_else(|| {
                        anyhow!("PDA Blob account path is empty (should have at least the account name)")
                    })?;
                let instruction_account_name =
                    instruction_account_name.key().context("Account name")?;
                if account_content_path.is_empty() {
                    return idl_map_get_key_or_else(
                        instruction_addresses,
                        instruction_account_name,
                    )
                    .context("Instruction addresses")
                    .map(|address| address.to_bytes().to_vec());
                }
                let instruction_account_state = idl_map_get_key_or_else(
                    instruction_accounts_states,
                    instruction_account_name,
                )
                .context("Instruction accounts states")?;
                let value = account_content_path
                    .try_get_json_value(instruction_account_state)
                    .context("Account extract value")?;
                if let Some(typing) = typing {
                    typing
                        .1
                        .try_serialize(value, &mut data, prefixed)
                        .context("Serialize Blob Bytes")?;
                } else {
                    let instruction_account_content_type_full =
                        instruction_accounts_contents_type_full
                            .get(instruction_account_name)
                            .context("Account content type")?;
                    account_content_path
                        .try_get_type_full(
                            instruction_account_content_type_full,
                        )
                        .context("Account content type from path")?
                        .try_serialize(value, &mut data, prefixed)
                        .context("Serialize Blob Bytes")?;
                };
            },
        };
        Ok(data)
    }
}
